phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,533 @@
+    // lower-numbered PCI device (4) and end up in Alpine 3.20.
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
+
+    let boot_num_bytes = read_efivar(&vm, BOOT_CURRENT_VAR).await?;
+
+    let boot_num: u16 = u16::from_le_bytes(boot_num_bytes.try_into().unwrap());
+
+    let boot_option_bytes = read_efivar(&vm, &bootvar(boot_num)).await?;
+
+    let mut cursor = Cursor::new(boot_option_bytes.as_slice());
+
+    let load_option = EfiLoadOption::parse_from(&mut cursor).unwrap();
+
+    assert!(load_option.path.matches_pci_device_function(4, 0));
+
+    // Now specify a boot order and do the whole thing again. Note that this
+    // order puts the later PCI device first, so this changes the boot order!
+    cfg.boot_order(vec!["alt-boot", "boot-disk"]);
+
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
+
+    let boot_num_bytes = read_efivar(&vm, BOOT_CURRENT_VAR).await?;
+
+    let boot_num: u16 = u16::from_le_bytes(boot_num_bytes.try_into().unwrap());
+
+    let boot_option_bytes = read_efivar(&vm, &bootvar(boot_num)).await?;
+
+    let mut cursor = Cursor::new(boot_option_bytes.as_slice());
+
+    let load_option = EfiLoadOption::parse_from(&mut cursor).unwrap();
+
+    // If we were going to test the PCI bus number too, we'd check the AHCI
+    // Device Path entry that precedes these PCI values. But we only use PCI bus
+    // 0 today, and the mapping from an AHCI Device Path to a PCI root is not
+    // immediately obvious?
+    assert!(load_option.path.matches_pci_device_function(24, 0));
+}
+
+// This is very similar to the `in_memory_backend_smoke_test` test, but
+// specifically asserts that the unbootable disk is first in the boot order; the
+// system booting means that boot order is respected and a non-bootable disk
+// does not wedge startup.
+#[phd_testcase]
+async fn unbootable_disk_skipped(ctx: &Framework) {
+    let mut cfg = ctx.vm_config_builder("unbootable_disk_skipped");
+
+    cfg.data_disk(
+        "unbootable",
+        DiskSource::FatFilesystem(FatFilesystem::new()),
+        DiskInterface::Virtio,
+        DiskBackend::InMemory { readonly: true },
+        16,
+    );
+
+    // `boot-disk` is the implicitly-created boot disk made from the default
+    // guest OS artifact.
+    //
+    // explicitly boot from it later, so OVMF has to try and fail to boot
+    // `unbootable`.
+    cfg.boot_order(vec!["unbootable", "boot-disk"]);
+
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
+
+    let boot_num_bytes = read_efivar(&vm, BOOT_CURRENT_VAR).await?;
+
+    let boot_num: u16 = u16::from_le_bytes(boot_num_bytes.try_into().unwrap());
+
+    let boot_option_bytes = read_efivar(&vm, &bootvar(boot_num)).await?;
+
+    let mut cursor = Cursor::new(boot_option_bytes.as_slice());
+
+    let load_option = EfiLoadOption::parse_from(&mut cursor).unwrap();
```

Comment:
i wonder if this ought to be a function, 'cause it also shows up twice in the `configurable_boot_order` test?